spark/src/main/java/org/apache/iceberg/actions/ExpireSnapshotsAction.java

```diff
@@ -0,0 +1,260 @@
+  }
+
+  public ExpireSnapshotsAction expireSnapshotId(long expireSnapshotId) {
+    localExpireSnapshots.expireSnapshotId(expireSnapshotId);
+    return this;
+  }
+
+  public ExpireSnapshotsAction expireOlderThan(long timestampMillis) {
+    localExpireSnapshots.expireOlderThan(timestampMillis);
+    return this;
+  }
+
+  public ExpireSnapshotsAction retainLast(int numSnapshots) {
+    localExpireSnapshots.retainLast(numSnapshots);
+    return this;
+  }
+
+  public ExpireSnapshotsAction deleteWith(Consumer<String> newDeleteFunc) {
+    deleteFunc = newDeleteFunc;
+    return this;
+  }
+
+
+  @Override
+  protected Table table() {
+    return table;
+  }
+
+  /**
+   * Execute is a synonym for commit in this implementation. Calling either commit or execute will
+   * launch the Spark equivalent of RemoveSnapshots.
+   *
+   * @return nothing
+   */
+  @Override
+  public ExpireSnapshotResults execute() {
+    localExpireSnapshots.commit();
+
+    TableMetadata currentMetadata = ops.refresh();
+
+    //Locally determine which snapshots have been expired
+    SnapshotExpirationChanges expiredSnapshotChanges =
+        ExpireSnapshotUtil.getExpiredSnapshots(currentMetadata, base);
+
+    //Locally determine which manifests will need to be scanned, reverted, deleted
+    ManifestExpirationChanges manifestExpirationChanges =
+        ExpireSnapshotUtil.determineManifestChangesFromSnapshotExpiration(
+            expiredSnapshotChanges.validSnapshotIds(), expiredSnapshotChanges.expiredSnapshotIds(),
+            currentMetadata, base, ops.io());
+
+    FileIO io = SparkUtil.serializableFileIO(table);
+
+    //Going the RDD Route because our reader functions all work with full Manifest Files
+    JavaSparkContext javaSparkContext = JavaSparkContext.fromSparkContext(session.sparkContext());
+
+    JavaRDD<ManifestFile> manifestsToScan =
+        javaSparkContext
+            .parallelize(new LinkedList<>(manifestExpirationChanges.manifestsToScan()));
+
+    JavaRDD<ManifestFile> manifestsToRevert =
+        javaSparkContext
+            .parallelize(new LinkedList<>(manifestExpirationChanges.manifestsToRevert()));
+
+    FileIO serializableIO = SparkUtil.serializableFileIO(table);
+
+    Broadcast<Map<Integer, PartitionSpec>> broadcastedSpecLookup =
+        javaSparkContext.broadcast(ops.current().specsById());
+
+    Broadcast<Set<Long>> broadcastValidIDs =
+        javaSparkContext.broadcast(expiredSnapshotChanges.validSnapshotIds());
+
+    JavaRDD<String> filesToDeleteFromScan = manifestsToScan.mapPartitions(manifests -> {
+      Map<Integer, PartitionSpec> specLookup = broadcastedSpecLookup.getValue();
+      Set<Long> validIds = broadcastValidIDs.getValue();
+      Set<String> filesToDelete = new HashSet<>();
+      Tasks.foreach(ImmutableList.copyOf(manifests))
+          .retry(3).suppressFailureWhenFinished()
+          .executeWith(ThreadPools.getWorkerPool())
```

Comment:
I am not sure we need to use `ThreadPools` on executors. First, `ThreadPools` in Iceberg was meant to parallelize some work on the driver. Second, we will have only one core executing a task in most cases. Since we are already parallelizing the work using Spark, I don't think we need the thread pool here.

Reply:
I like to parallelize within executors as well just because it is more flexible that doing so with number of tasks but I can remove it if you want. In general I don't like doing any blocking IO if I can help it.